File: backend/src/main/java/com/shopmanagement/label/service/LabelTemplateService.java
package com.shopmanagement.label.service;

import com.shopmanagement.label.dto.LabelTemplateRequest;
import com.shopmanagement.label.dto.LabelTemplateResponse;
import com.shopmanagement.label.entity.LabelTemplate;
import com.shopmanagement.label.repository.LabelTemplateRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class LabelTemplateService {

    private final LabelTemplateRepository labelTemplateRepository;

    @Transactional(readOnly = true)
    public List<LabelTemplateResponse> getAll() {
        return labelTemplateRepository.findAllByOrderByUpdatedAtDesc().stream()
                .map(this::toResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public LabelTemplateResponse getById(Long id) {
        LabelTemplate template = labelTemplateRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Label template not found: " + id));
        return toResponse(template);
    }

    /** Returns the common/default template, or null if none has been saved yet. */
    @Transactional(readOnly = true)
    public LabelTemplateResponse getDefault() {
        return labelTemplateRepository.findFirstByIsDefaultTrueOrderByUpdatedAtDesc()
                .map(this::toResponse)
                .orElse(null);
    }

    public LabelTemplateResponse create(LabelTemplateRequest request) {
        String user = currentUser();
        LabelTemplate template = LabelTemplate.builder()
                .name(request.getName())
                .shopId(request.getShopId())
                .labelWidthMm(request.getLabelWidthMm())
                .labelHeightMm(request.getLabelHeightMm())
                .isDefault(request.getIsDefault() != null ? request.getIsDefault() : Boolean.TRUE)
                .design(request.getDesign())
                .createdBy(user)
                .updatedBy(user)
                .build();
        template = labelTemplateRepository.save(template);
        log.info("Created label template id={} by {}", template.getId(), user);
        return toResponse(template);
    }

    public LabelTemplateResponse update(Long id, LabelTemplateRequest request) {
        LabelTemplate template = labelTemplateRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Label template not found: " + id));
        template.setName(request.getName());
        template.setShopId(request.getShopId());
        template.setLabelWidthMm(request.getLabelWidthMm());
        template.setLabelHeightMm(request.getLabelHeightMm());
        if (request.getIsDefault() != null) {
            template.setIsDefault(request.getIsDefault());
        }
        template.setDesign(request.getDesign());
        template.setUpdatedBy(currentUser());
        template = labelTemplateRepository.save(template);
        log.info("Updated label template id={}", id);
        return toResponse(template);
    }

    /**
     * Upsert for the single "common" template: updates the existing default if one
     * exists, otherwise creates it. This backs the designer's single Save button.
     */
    public LabelTemplateResponse saveDefault(LabelTemplateRequest request) {
        request.setIsDefault(Boolean.TRUE);
        return labelTemplateRepository.findFirstByIsDefaultTrueOrderByUpdatedAtDesc()
                .map(existing -> update(existing.getId(), request))
                .orElseGet(() -> create(request));
    }

    public void delete(Long id) {
        if (!labelTemplateRepository.existsById(id)) {
            throw new IllegalArgumentException("Label template not found: " + id);
        }
        labelTemplateRepository.deleteById(id);
        log.info("Deleted label template id={}", id);
    }

    private String currentUser() {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        return auth != null ? auth.getName() : "system";
    }

    private LabelTemplateResponse toResponse(LabelTemplate t) {
        return LabelTemplateResponse.builder()
                .id(t.getId())
                .name(t.getName())
                .shopId(t.getShopId())
                .labelWidthMm(t.getLabelWidthMm())
                .labelHeightMm(t.getLabelHeightMm())
                .isDefault(t.getIsDefault())
                .design(t.getDesign())
                .createdBy(t.getCreatedBy())
                .updatedBy(t.getUpdatedBy())
                .createdAt(t.getCreatedAt())
                .updatedAt(t.getUpdatedAt())
                .build();
    }
}

File: frontend/src/app/core/services/label-template.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../models/api-response.model';
import { LabelTemplate } from '../models/label-template.model';

@Injectable({ providedIn: 'root' })
export class LabelTemplateService {
  private readonly API_URL = `${environment.apiUrl}/label-templates`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<LabelTemplate[]> {
    return this.http.get<ApiResponse<LabelTemplate[]>>(this.API_URL)
      .pipe(map(res => res.data || []));
  }

  /** The common/default template (null if none saved yet). */
  getDefault(): Observable<LabelTemplate | null> {
    return this.http.get<ApiResponse<LabelTemplate>>(`${this.API_URL}/default`)
      .pipe(map(res => res.data || null));
  }

  getById(id: number): Observable<LabelTemplate> {
    return this.http.get<ApiResponse<LabelTemplate>>(`${this.API_URL}/${id}`)
      .pipe(map(res => res.data));
  }

  /** Upsert the single common template. */
  saveDefault(template: LabelTemplate): Observable<LabelTemplate> {
    return this.http.post<ApiResponse<LabelTemplate>>(`${this.API_URL}/default`, template)
      .pipe(map(res => res.data));
  }

  delete(id: number): Observable<void> {
    return this.http.delete<ApiResponse<void>>(`${this.API_URL}/${id}`)
      .pipe(map(() => void 0));
  }
}
